### DISCUSSION TITLE
How can I use global variables in a plugin?

### DISCUSSION THREAD

**Discussion by s22-tech** [YOURLS/YOURLS#3467] [Plugins]
I'd like to create some variables at the beginning of a plugin and then use then in my functions.  But calling `global $var;` at the start of a function does nothing - the $var is not available to the function.

If I define a CONSTANT, I can use that in the functions but I'm curious as to why a global $var isn't useable?

(I know some people don't like globals, but I don't mind them in a situation like this.)

**Comment by dgw** [comment]
Need example failing code, probably. Per [PHP docs](https://www.php.net/manual/en/language.variables.scope.php) about scope, using `global $varname` should indeed just work.

**Reply by s22-tech** [reply]
Here's an example.  Notice how the variable $xxx returns null in the output:

```
<?php
/*
Plugin Name: API Action - Locate keyword
Plugin URI: https://github.com/s22-tech/
Description: Custom API action to see if keyword exists
Version: 1.2
Author: s22_tech
Author URI: https://github.com/s22-tech/
Date: 2021-01-03
Based On: https://github.com/YOURLS/API-action
Returns: json string
*/


$xxx = 'zzzzz';

yourls_add_filter('api_action_locate_keyword', 's22_locate_keyword');

function s22_locate_keyword () {
	global $xxx;

	if (!isset($_REQUEST['keyword'])) {
		return [
			'statusCode' => 400,
			'simple'     => "Need a 'keyword' parameter",
			'message'    => 'error: missing param',
		];
	}

	$keyword = $_REQUEST['keyword'];


	if (yourls_is_shorturl($keyword)) {
		return [
			'status_code' => 200,
			'status'      => 'success',
			'message'     => 'keyword found',
			'keyword'     => $keyword,
			'xxx' => $xxx,
		];
	}
	else {
		return [
			'status_code' => 404,
			'status'      => 'failed',
			'message'     => 'error: not found',
			'keyword'     => '',
			'xxx' => $xxx,
		];
	}
}

/*

Example URL:
https://example.com/yourls-api.php?signaturexxxx&action=locate_keyword&keyword=abc101&format=json

*/
```

**Reply by dgw** [reply]
If that is essentially the real code you are trying to get working, that exact functionality is already available via the built-in `action=expand`. (Well, OK, it doesn't have the mystery `xxx` parameter. But I still thought I'd mention it in case it can save you plugin-dev time.)

<details>
<summary>Example <tt>expand</tt> API calls</summary>

**Valid keyword**, e.g. `https://yourls.org/yourls-api.php?signature=<token>&action=expand&shorturl=knownkw&format=json`:

```json
{
    "keyword": "knownkw",
    "shorturl": "https:\/\/yourls.org\/knownkw",
    "longurl": "https:\/\/www.google.com\/search?q=turlututu",
    "title": "https:\/\/www.google.com\/search?q=turlututu",
    "message": "success",
    "statusCode": 200
}
```

(served with HTTP 200 status code)

----

**Nonexistent keyword**, e.g. `https://yourls.org/yourls-api.php?signature=<token>&action=expand&shorturl=fakekw&format=json`:

```json
{
    "keyword": "fakekw",
    "message": "Error: short URL not found",
    "errorCode": 404
}
```

(served with HTTP 404 status code)

</details>

All of that said, it works if you declare `global $xxx;` in the top-level scope, before setting `$xxx = 'zzzzz';`.

```php
/*
  (plugin header, omitted for brevity)
*/

global $xxx;

$xxx = 'zzzzz';

yourls_add_filter('api_action_locate_keyword', 's22_locate_keyword');

// rest of plugin code
```

**Reply by s22-tech** [reply]
Hmm... that works, but why would you have to declare the $var as global twice in the plugin?

Also, could you expound on `action=expand` and perhaps give an example?

**Reply by dgw** [reply]
> why would you have to declare the $var as global twice in the plugin?

You probably don't need to declare it `global` inside the function if you're only reading the value; I just didn't bother testing any more after I got it working. My primary language is Python, not PHP 😅

> could you expound on `action=expand` and perhaps give an example?

Expand the "Example `expand` API calls` line in my previous comment.

**Reply by s22-tech** [reply]
> You probably don't need to declare it global inside the function

Unfortunately, you do.  Without it you'll get an error.  But that's not how it's supposed to work.  I've never had to use, nor even seen the use of, multiple global calls like that.

**Reply by dgw** [reply]
> But that's not how it's supposed to work. I've never had to use, nor even seen the use of, multiple global calls like that.

It is probably to do with how YOURLS loads plugins. As I said, I'm not super plugged into how PHP works, but I did see some discussions about how scoping can work in unexpected ways when files are `include`d/`require`d. That's as far as I could go with it rn.

_I had started responding to what you said about `action=expand`, but now that you've deleted that part of your comment I'm guessing you figured it out. 🙂_

**Reply by s22-tech** [reply]
> It is probably to do with how YOURLS loads plugins.

That's unfortunate.  I'll stick with declaring constants for now, unless someone here has any ideas.

> I'm guessing you figured it out.

Well, I guessed you were talking about replacing `action=locate_keyword` with `action=expand` but doing that returns the wrong information, e.g. `{"keyword":"","message":"Error: short URL not found","errorCode":404}` when testing for a known good keyword.

So, unless I'm using it incorrectly, it's not the same.

**Reply by dgw** [reply]
`action=expand&shorturl=kwhere`

It does not take exactly the same parameter as your custom `action=locate_keyword`, which takes `keyword=kwhere` instead of `shorturl=kwhere`

**Comment by s22-tech** [comment]
Regarding the `action=expand` keyword you mentioned here, I'd like to suggest that it return consistent keys for both success and failure.  

On success, it returns 'keyword', 'shorturl', 'longurl', 'title', 'message', and 'statusCode'.

But on failure, it returns 'keyword', 'message', and 'errorCode'.

Shouldn't 'errorCode' be named 'statusCode' so that we can test for a single keyword?
